pkg/agent/secrets.go

```diff
@@ -0,0 +1,159 @@
+package agent
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	apiv1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+type secretSyncAction string
+
+const (
+	secretSyncActionSet    = secretSyncAction("SET")
+	secretSyncActionDelete = secretSyncAction("DELETE")
+)
+
+type secretInterface interface {
+	Create(ctx context.Context, secret *apiv1.Secret, opts metav1.CreateOptions) (*apiv1.Secret, error)
+	Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
+	Get(ctx context.Context, name string, opts metav1.GetOptions) (*apiv1.Secret, error)
+	Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *apiv1.Secret, err error)
+}
+
+// secretsGetterFactory is a factory for creating SecretsGetter.
+type secretsGetterFactory func(namespace string) (secretInterface, error)
+
+type secretSyncCommand struct {
+	name      string
+	namespace string
+	action    secretSyncAction
+	secret    map[string][]byte
+}
+
+func (s *secretSyncCommand) String() string {
+	return fmt.Sprintf("<secret=%s namespace=%s action=%s>", s.name, s.namespace, s.action)
+}
+
+func (s *secretSyncCommand) RunWithClientFactory(
+	ctx context.Context, secretGetterFactory secretsGetterFactory,
+) error {
+	client, err := secretGetterFactory(s.namespace)
+	if err != nil {
+		return err
+	}
+
+	return s.syncSecret(ctx, client)
+}
+
+func (s *secretSyncCommand) getOps() (ops []map[string]string) {
+	switch s.action {
+	case secretSyncActionDelete:
+		for key := range s.secret {
+			ops = append(ops, map[string]string{
+				"op":   "remove",
+				"path": fmt.Sprintf("/data/%s", key),
+			})
+		}
+	default:
```

Comment:
Since this is using an alias on a string there is nothing stopping someone from creating a `secretSyncCommannd` with action "foo". 

Therefore, we should explicitly catch the case `secretSyncActionSet` and then the default case should probably `panic` because that would mean we have a programming error somewhere and we would want to catch that in CI.